Replace read/write lock in JarResource to avoid virtual threads pinning

mariofusco opened the pull request:
This pull request fixes the [race condition](https://github.com/quarkusio/quarkus/issues/42067) present in [this commit](https://github.com/quarkusio/quarkus/commit/ff4b32bb0d8a1df0ccf1cc86a1e1ae9a840c5b0f) that, because of it, had to be reverted.

/cc @geoand @gsmet @franz1981 @jedla97 @michalvavrik 

Fixes: #42067

The nature of the fix is described at https://github.com/quarkusio/quarkus/pull/42139#pullrequestreview-2199296948

mariofusco committed:
Replace read/write lock in JarResource to avoid virtual threads pinning

add test

format test

recompile test jars with jdk17

mariofusco commented on independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java:
```diff
@@ -0,0 +1,176 @@
+
+    private static JarFileReference asyncLoadAcquiredJarFile(JarResource jarResource) {
+        CompletableFuture<JarFileReference> newJarRefFuture = new CompletableFuture<>();
+        CompletableFuture<JarFileReference> existingJarRefFuture = null;
```
The introduction of this variable is the actual fix. The old algorithm [wrongly reused](https://github.com/quarkusio/quarkus/commit/ff4b32bb0d8a1df0ccf1cc86a1e1ae9a840c5b0f#diff-f20997e90aae1d2fe046b512defb1a3e14215b7a834177d43369542bae99a7f6R168) the `newJarFileRef` one. This means that in the unfortunate case when a thread was [about to close the old jar](https://github.com/quarkusio/quarkus/commit/ff4b32bb0d8a1df0ccf1cc86a1e1ae9a840c5b0f#diff-f20997e90aae1d2fe046b512defb1a3e14215b7a834177d43369542bae99a7f6R60), but [before setting the AtomicReference to null](https://github.com/quarkusio/quarkus/commit/ff4b32bb0d8a1df0ccf1cc86a1e1ae9a840c5b0f#diff-f20997e90aae1d2fe046b512defb1a3e14215b7a834177d43369542bae99a7f6R67), another thread could fail the `compareAndSet(null)` check here, take at the end of the loop from the `AtomicReference` the old and no longer valid `CompletableFuture` and in subsequent loop (after that the close finally managed to set the `AtomicReference` to null) mistakenly set it back into the same `AtomicReference`.

geoand commented:
Thanks for the fix @mariofusco !

@jedla97 thanks for signing up to test the fix! Hopefully @famod can do so as well

jedla97 commented:
@mariofusco Big thanks for the fix as I believe it was hard to find it. I tested it (run two reproducer multiple times) and the race condition didn't appear. So from my POV the original issue is fixed (hopefully we won't see any different :pray: ).

franz1981 commented:
It would be great if there is a test which reproduce it too, with the changes to fix

famod commented:
Sorry, I am unable to test this ATM. On my work machine, I'm getting a puzzling dependency resolution issue while building, it's related to gradle-tooling.

mariofusco commented:
> It would be great if there is a test which reproduce it too, with the changes to fix

I did a few more checks on this fix and also added a proper unit test. The unit test reproduces exactly the same situation that caused the `IllegalStateException` in the former version of this class loader as I described it [here](https://github.com/quarkusio/quarkus/pull/42139#discussion_r1691438739). Of course, being a race condition, the test is not deterministic and it fails with the older version extremely seldomly. However it's descriptive of the problem and I'm afraid it will be impossible to make it reproducible in a more reliable way. 

@geoand @franz1981 the pull request has been already rebased and squashed as usual, so it should be ready to be merged. Please let me know if this is enough or you see any room for further improvements.

franz1981 reviewed (approved):
LGTM well done as usual Mario!

geoand reviewed (approved):
Thanks a lot!

mariofusco commented:
> ## Failing Jobs
> Status 	Name 	Step 	Failures 	Logs 	Raw logs 	Build scan
> ✖ 	JVM Tests - JDK 17 	`Build` 	[Failures](#user-content-test-failures-job-28064977823) 	[Logs](https://github.com/quarkusio/quarkus/actions/runs/10149362227/job/28064977823) 	[Raw logs](https://github.com/quarkusio/quarkus/commit/b2b793c9e7bf6d5cebe860fe556af2e2aa5a094c/checks/28064977823/logs) 	[🔍](https://ge.quarkus.io/s/nvcms2tcq3nba)
> ✖ 	JVM Tests - JDK 17 Windows 	`Build` 	[Failures](#user-content-test-failures-job-28064978975) 	[Logs](https://github.com/quarkusio/quarkus/actions/runs/10149362227/job/28064978975) 	[Raw logs](https://github.com/quarkusio/quarkus/commit/b2b793c9e7bf6d5cebe860fe556af2e2aa5a094c/checks/28064978975/logs) 	[🔍](https://ge.quarkus.io/s/2e7oemre4jm26)

These failures are caused by the fact that I compiled the minimal jars that I used for this test with Java 21, so they cannot be loaded with Java 17.

`java.lang.UnsupportedClassVersionError: org/simple/SimplePojo1 has been compiled by a more recent version of the Java Runtime (class file version 65.0), this version of the Java Runtime only recognizes class file versions up to 61.0`

I will recompile those jars with Java 17 and replace them in this pull request immediately.

mariofusco commented:
> I will recompile those jars with Java 17 and replace them in this pull request immediately.

Done.

mariofusco commented:
Is this ready to be merged? Since the 3.13.0 is already out there shouldn't be any reason to hold it, right?

geoand commented:
Correct

geoand commented:
Let's keep an eye out on weird failures as we did with the previous iteration.

mariofusco commented:
> Let's keep an eye out on weird failures as we did with the previous iteration.

Thank you for merging. Sure, please ping me as soon as you will notice something wrong.

geoand commented:
Will do! Thanks again for tackling this!

If anything comes up, it will be by QE running their testsuite over the next few days. Otherwise if there is an issue, reports from users will start coming in when `3.14.0.CR1` is released.
